### PR TITLE
Remove kernel support for futexes

### PR DESCRIPTION
**Pull request by jermar**
This PR removes futexes from the kernel and replaces them with a much leaner mechanism based on wait queue made usable from uspace via capabilities. As stated in a554ae8, the replacement is not meant to be used between multiple address spaces. For that, one would have to separate the futex counter and the wait queue handle used by the futex. The counter would need to be shared with the other address space and the waitq capability would have to be preallocated and passed to the other task (which we cannot do yet).

### CODE REVIEW

**Comment by jermar** [comment]
I tested this with multithreading enabled in VFS. The system runs fine, but after it allocates the first waitq (as can be observed in kconsole by typing caches), some things start to get slower, such as printing out the list of caches. I will keep looking into this.

File: uspace/lib/c/generic/private/futex.h [hunk 1]
```diff
@@ -146,7 +169,7 @@ static inline errno_t futex_down_composable(futex_t *futex,
 static inline errno_t futex_up(futex_t *futex)
 {
 	if (atomic_fetch_add_explicit(&futex->val, 1, memory_order_release) < 0)
-		return __SYSCALL1(SYS_FUTEX_WAKEUP, (sysarg_t) futex);
```
Comment (jermar) [inline uspace/lib/c/generic/private/futex.h]: Actually, this is not correct, the wakeup might need to allocate the handle here too. But what to do with a failure to allocate it? Going to switch to the allocate-before-decrement in sleep strategy.

File: kernel/generic/src/synch/syswaitq.c [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+/** Initialize the user waitq subsystem */
+void sys_waitq_init(void)
+{
+	waitq_cache = slab_cache_create("waitq_t", sizeof(waitq_t), 0, NULL,
+	    NULL, 0);
+}
+
+/** Clean-up all waitq capabilities held by the exiting task */
+void sys_waitq_task_cleanup(void)
```
Comment (le-jzr) [inline kernel/generic/src/synch/syswaitq.c]: So, why do we need a separate cleanup routine for each kobject type, instead of just one generic cleanup?

Comment (jermar) [reply kernel/generic/src/synch/syswaitq.c]: Slightly different semantics of the objects. Sometimes you need to do further actions. This should eventually become normalized, but not now.

File: uspace/lib/c/generic/thread/fibril.c [hunk 3]
```diff
@@ -795,7 +795,8 @@ int fibril_test_spawn_runners(int n)
 
 	if (!multithreaded) {
 		_ready_debug_check();
-		futex_initialize(&ready_semaphore, ready_st_count);
+		if (futex_initialize(&ready_semaphore, ready_st_count) != EOK)
+			abort();
```
Comment (le-jzr) [inline uspace/lib/c/generic/thread/fibril.c]: Random idea: What if instead of aborting on failure, we fall back to a spinlock+thread_yield()?

**Comment by jxsvoboda** [comment]
LGTM

**Pull request merged**